### PR TITLE
[Tests fail `iris>=3.2.0`] Add a test module that runs tests of `iris` I/O everytime we notice serious bugs there

### PR DESCRIPTION
**Pull request by valeriupredoi**
## Description

@schlunma finding the serious saving issue presented in #1509 prompted me to think we should have a test module that tests all these serious bugs from `iris` IO, no matter they fix them - just so that we are sure they don't resurface. These are cheap tests and shouldn't be a burden on the testing and CI performance.

- Test runs with `iris=3.2.0.post0` FAIL [here](https://github.com/ESMValGroup/ESMValCore/runs/5321735240?check_suite_focus=true)
- Test runs with `iris=3.1` PASS [here](https://github.com/ESMValGroup/ESMValCore/runs/5360398116?check_suite_focus=true)

***

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by zklaus** [comment]
I like the idea. Out of curiosity: Does anything change if the cubes data and dim coord are also lazy?

**Comment by valeriupredoi** [comment]
@schlunma have a look at this test - it shows a new facet of the issue you detail in https://github.com/SciTools/iris/issues/4599 ie the saving happens fine in a pytest environment when run individually with `pytest tests/unit/test_iris_io.py` (NOT saving well in a GA or Circle test suite though), but the loading fails:
```
========================================================== FAILURES ==========================================================
____________________________________________ test_iris_save_with_lazy_coordinate _____________________________________________

tmp_path = PosixPath('/tmp/pytest-of-valeriu/pytest-189/test_iris_save_with_lazy_coord0')

    def test_iris_save_with_lazy_coordinate(tmp_path):
        """
        Test saving a cube with fully lazy coords.
    
        Motivated by https://github.com/SciTools/iris/issues/4599
        """
        print("iris version:", iris.__version__)
        cube = create_fully_lazy_cube()
        save_path = tmp_path / 'test_iris_v32.nc'
        iris.save(cube, save_path)
        print("Attempted to load ", save_path)
>       loaded_cube = iris.load_cube(save_path)

tests/unit/test_iris_io.py:37: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../miniconda3/envs/rc3-3/lib/python3.10/site-packages/iris/__init__.py:343: in load_cube
    cube = cubes.merge_cube()
../miniconda3/envs/rc3-3/lib/python3.10/site-packages/iris/cube.py:394: in merge_cube
    (merged_cube,) = proto_cube.merge()
../miniconda3/envs/rc3-3/lib/python3.10/site-packages/iris/_merge.py:1250: in merge
    indexes = build_indexes(positions)
../miniconda3/envs/rc3-3/lib/python3.10/site-packages/iris/_merge.py:606: in build_indexes
    name_index_by_scalar[value] = {
../miniconda3/envs/rc3-3/lib/python3.10/site-packages/iris/_merge.py:607: in <dictcomp>
    other_name: set((position[other_name],))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Cell(point=4, bound=(0, masked))

    def __hash__(self):
>       return super().__hash__()
E       TypeError: unhashable type: 'MaskedConstant'

../miniconda3/envs/rc3-3/lib/python3.10/site-packages/iris/coords.py:1347: TypeError
---------------------------------------------------- Captured stdout call ----------------------------------------------------
iris version: 3.2.0.post0
Attempted to load  /tmp/pytest-of-valeriu/pytest-189/test_iris_save_with_lazy_coord0/test_iris_v32.nc
====================================================== warnings summary ======================================================
```

**Comment by valeriupredoi** [comment]
> I like the idea. Out of curiosity: Does anything change if the cubes data and dim coord are also lazy?

just made the data fully lazy from get going too, fails the same way :+1:

**Comment by valeriupredoi** [comment]
OK so I expanded on the test cases: fully lazy (coords + data) cube fails to save correctly - the important bit here is the lazy coord, normal numpy array (coords + data) saves nicely

File: tests/unit/test_iris_io.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def test_iris_save_with_lazy_coordinate(tmp_path):
     save_path = tmp_path / 'test_iris_v32.nc'
     iris.save(cube, save_path)
     print("Attempted to load ", save_path)
-    loaded_cube = iris.load_cube(save_path)
+    loaded_cube = iris.load_cube(str(save_path))
```
Comment (zklaus) [inline tests/unit/test_iris_io.py:54]: I think the preferred method here would be `save_path.as_posix()`.

**Comment by schlunma** [comment]
Cheers, thanks V.!! Please merge or approve when you are happy with this @zklaus

**Review by zklaus — APPROVED** [review]

**Pull request merged**